### PR TITLE
First pass refactor of case resource creation flow

### PR DESCRIPTION
**Pull request by wssheldon**

### CODE REVIEW

**Comment by wssheldon** [comment]
I'm not sure what type `individual_participants` is. It is `list[str]` in `main` but we call `.email` on it.

File: src/dispatch/case/flows.py [hunk 1]
```diff
@@ -608,107 +610,178 @@ def case_assign_role_flow(
...
+        self.update_ticket()
+
+    @check_resource_exists("groups")
+    def create_group(
+        self,
+        individual_participants: ParticipantEmails,
+        team_participants: ParticipantEmails,
+    ) -> None:
+        """Create a new group for the case if it does not exist."""
+        participant_emails = set(individual_participants) | {t.email for t in team_participants}
```
Comment (mvilanova) [inline src/dispatch/case/flows.py:692]: Should we `raise AlreadyExists` if it exists?

File: src/dispatch/case/flows.py [hunk 2]
```diff
@@ -608,107 +610,178 @@ def case_assign_role_flow(
...
-        storage_members = []
-        if case.tactical_group:
-            storage_members = [case.tactical_group.email]
-        # direct add members if not group exists
-        else:
-            storage_members = direct_participant_emails
+    @check_resource_exists("storage")
+    def create_storage(self, individual_participants: ParticipantEmails) -> None:
+        """Create a storage resource for the case if it does not exist."""
+        storage_members = [p.email for p in individual_participants]
```
Comment (mvilanova) [inline src/dispatch/case/flows.py:703]: Should we `raise AlreadyExists` if it exists?

File: src/dispatch/case/flows.py [hunk 3]
```diff
@@ -608,107 +610,178 @@ def case_assign_role_flow(
...
-            description="Conversation added to case",
-            case_id=case.id,
-        )
-        # wait until all resources are created before adding suggested participants
-        individual_participants = [x.email for x, _ in individual_participants]
+    @check_resource_exists("conversation")
+    def create_conversation(
+        self, conversation_target: str | None, individual_participants: ParticipantEmails
+    ) -> None:
+        """Create a conversation thread for the case and add participants."""
```
Comment (mvilanova) [inline src/dispatch/case/flows.py:737]: Should we `raise AlreadyExists` if it exists?

**Review by mvilanova — APPROVED** [review]

File: src/dispatch/case/flows.py (hunk 3 above)
Comment (wssheldon) [reply src/dispatch/case/flows.py:737]: The decorator raises that error, if you're referring to the resource:

https://github.com/Netflix/dispatch/pull/4164/files#diff-ee560a5825ccb96768d929d3eb4a3cc23526ce20e31155e79dda5180e9065f0bR618

**Pull request closed without merging**